### DISCUSSION TITLE
[Question] Manually updatable progress bars

### DISCUSSION THREAD

**Discussion by andrzejnovak** [Textualize/rich#399] [General] [answered]
Hi, first let me just say this lib looks awesome. I am currently using something else to render a histogram like this to command line and was trying in vain to animate it.

![image](https://user-images.githubusercontent.com/13226500/96796764-09172780-1400-11eb-924b-684d43cb45da.png)

It looks to me like I could use something like the multi progress bar example, but looking at the code it was not immediately obvious to my how to do this. Each bin would be a "task" that I could then update. 

I suppose I'd have to create subclass of BarColumn to change the looks and somehow plug it into the Progress class?

Could you point me in the right direction?

**Comment by willmcgugan** [comment]
You *could* render this with the Progress class, by having a task for each data point, but I suspect it would be simpler for you to generate a Table with a row for labels and a row for bars.

Start by creating a grid (table with no lines):

```python
my_table = Table.grid()
my_table.title = "Samples"
```

Then add rows consisting of the label and a Bar.

```python
for i, x in enumerate(data):
    my_table.add_row(str(i), Bar(size=100, end=y)) 
```

As for animation, the simplest solution would be to clear the screen and redraw. Something along these lines:

```python
while True:
    with console:
        console.clear()
        console.print(my_table)
    update_data()
```

See examples/table_movie.py for an example of this kind of animation.

**Reply by andrzejnovak** [reply]
So the neat thing about the progress bars was not taking over the whole screen which the table_movie seems to do and I don't want. That's something I could probably do with curses or maybe just screen. 

I suppose my question is how to subclass the progress components, such that I could add tasks for each bin and tweak the line styling. I figured out how to set some of the colors, but was not able to change the width of the progress bar and the color of the "shadow" into which the progress line moves.

**Comment by willmcgugan** [comment, marked as answer]
The BarColumn accepts `bar_width`, which you can set to your desired width, or `None` for maximum available width.

You can also set `style` for the *shadow* color. See the [BarColumn reference](https://rich.readthedocs.io/en/latest/reference/progress.html#rich.progress.BarColumn) for details. You can subclass BarColumn or any other column class.

If you don't want to clear the entire screen, you can use `rich.live_render` which is what Progress uses under the hood. It's not documented yet, but it's straight forward to use.

**Comment by andrzejnovak** [comment]
Awesome, it worked in the end just using the base progress class.

![](https://i.imgur.com/KRneQOY.gif)

The only quirks are the flickering which seems to happen from the fifth line onwards and I am not entirely sure the background style is behaving correctly. I got the right look by passing `Style(color='black')`, but that's probably not general. I would expect `Style(color=None)` to render invisible aka the background should have the terminal colour, but instead, it ends up white.

**Reply by willmcgugan** [reply]
`Style(color=None)` doesn't mean no background I'm afraid. It means "background not set" so you'll get the default foreground color. Your solution will work as long as your terminal is dark.

Not sure where the flickering is coming from. If you could post rtest.py maybe I might be able to suggest some fixes.

**Reply by andrzejnovak** [reply]
Is there a way to set no background then? Or fetch the "current" one?

Sure, it's a tiny snippet:

<details>

```
import time
from rich.progress import Progress, BarColumn, TextColumn

from rich.style import Style
sty_bkg = Style(color='black')

import numpy as np 
H, E = [], []
for i in range(250):
    bins = np.linspace(0, 10, 40)
    h, e = np.histogram(np.random.normal(5, 2, 50), bins=bins)
    H.append(h)
    E.append(e)

with Progress(TextColumn("[bold blue]{task.description}", justify="right"), BarColumn(bar_width=200, style=sty_bkg)) as progress:

    max_width = 1000
    tasks = []
    for start in E[0]: # bins
        tasks.append(progress.add_task(str(round(start,2)), total=max_width))

    track_H = H[0]
    for i in range(250):
        if i > 0:
            track_H += H[i]
        if max(track_H) >= 0.9*max_width:
                max_width *= 1.5
        for value, task in zip(H[i], tasks):
            progress.update(task, advance=value, total=max_width)
        
        time.sleep(0.02)

```

</details>

I'm on WSL so I tested it with MS Terminal, hyper and the default ubuntu app and all seem to behave the same.  Strangely enough with only 6 "bins" there is no flicker.

**Comment by willmcgugan** [comment]
> Is there a way to set no background then? Or fetch the "current" one?

Afraid not. The shadow bar is drawn with a character, and there is no way to tell the terminal that the foreground color should be the same as the background. There's also no way to query it. It would be fairly simple to modify the ProgressBar class to not have a background.

I see a hint of flickering on MacOS iTerm, but nothing on Hyper or Terminal. I don't see any flickering on MS Terminal. So not sure why you get so much flickering.

You could try setting auto_refresh to False, and call progress.update() manually, once per loop.

**Reply by andrzejnovak** [reply]
Alas, that doesn't seem to change anything for me. But anyway, the use case I have in mind is likely to not update as fast as this demo, so it should still be ok.  Thanks for the feedback, the library is awesome 🙂
